src/main/scala/swiss/dasch/domain/BulkIngestService.scala

```diff
@@ -40,13 +22,30 @@ object IngestResult {
   val failed: IngestResult  = IngestResult(failed = 1)
 }
 
-final case class BulkIngestServiceLive(
+final case class BulkIngestService(
   storage: StorageService,
   ingestService: IngestService,
   config: IngestConfig,
-) extends BulkIngestService {
+  semaphoresPerProject: TMap[ProjectShortcode, TSemaphore],
+) {
+
+  private def withSemaphoreFork[E, A](
+    key: ProjectShortcode,
+    zio: ProjectShortcode => IO[E, A],
+  ): IO[Option[Nothing], Fiber.Runtime[E, A]] = {
+    def getSemaphore =
+      semaphoresPerProject.getOrElseSTM(key, TSemaphore.make(1)).tap(semaphoresPerProject.put(key, _)).commit
+    def acquireWithTimeout(sem: TSemaphore) =
+      sem.acquire.as(sem).commit.timeout(Duration.fromMillis(400)).some
+    getSemaphore
+      .flatMap(acquireWithTimeout)
+      .flatMap(sem => zio.apply(key).logError.ensuring(sem.release.commit).forkDaemon)
```

Comment:
The definitions here are a bit entangled, I suppose a for comprehension would've looked better here.